modyn/storage/include/internal/grpc/storage_service_impl.hpp

```diff
@@ -186,6 +186,136 @@ class StorageServiceImpl final : public modyn::storage::Storage::Service {
     return {StatusCode::OK, "Data retrieved."};
   }
 
+  template <typename WriterT>
+  Status GetDataPerWorker_Impl(  // NOLINT (readability-identifier-naming)
+      ServerContext* /*context*/, const modyn::storage::GetDataPerWorkerRequest* request, WriterT* writer) {
+    soci::session session = storage_database_connection_.get_session();
+    try {
+      // Check if the dataset exists
+      const int64_t dataset_id = get_dataset_id(session, request->dataset_id());
+
+      if (dataset_id == -1) {
+        SPDLOG_ERROR("Dataset {} does not exist.", request->dataset_id());
+        session.close();
+        return {StatusCode::OK, "Dataset does not exist."};
+      }
+
+      const int64_t start_timestamp = request->start_timestamp();
+      const int64_t end_timestamp = request->end_timestamp();
+
+      SPDLOG_INFO(
+          fmt::format("Received GetDataPerWorker Request for dataset {} (id = {}) and worker {} out of {} workers with"
+                      " start = {} and end = {}.",
+                      request->dataset_id(), dataset_id, request->worker_id(), request->total_workers(),
+                      start_timestamp, end_timestamp));
+
+      const int64_t total_keys =
+          get_number_of_samples_in_dataset_with_range(dataset_id, session, start_timestamp, end_timestamp);
+
+      if (total_keys > 0) {
+        int64_t start_index = 0;
+        int64_t limit = 0;
+        std::tie(start_index, limit) =
+            get_partition_for_worker(request->worker_id(), request->total_workers(), total_keys);
+
+        std::string query;
+        if (start_timestamp == 0 && end_timestamp == 0) {
+          query =
+              fmt::format("SELECT sample_id FROM samples WHERE dataset_id = {} ORDER BY sample_id LIMIT {} OFFSET {}",
+                          dataset_id, limit, start_index);
+        } else {
+          std::string timestamp_filter;
+          if (start_timestamp > 0 && end_timestamp == 0) {
+            timestamp_filter = fmt::format("updated_at >= {}", start_timestamp);
+          } else if (start_timestamp == 0 && end_timestamp > 0) {
+            timestamp_filter = fmt::format("updated_at < {}", end_timestamp);
+          } else if (start_timestamp > 0 && end_timestamp > 0) {
+            timestamp_filter = fmt::format("updated_at >= {} AND updated_at < {}", start_timestamp, end_timestamp);
+          } else if (start_timestamp == 0 && end_timestamp == 0) {
+            timestamp_filter = "1 = 1";
+          } else {
+            FAIL(fmt::format("Invalid timestamps: start = {}, end = {}", start_timestamp, end_timestamp));
+          }
```

Comment:
There is quite some code duplication here with the counting function. Could we extract this into a common function?

Reply:
Thanks for spotting this out!